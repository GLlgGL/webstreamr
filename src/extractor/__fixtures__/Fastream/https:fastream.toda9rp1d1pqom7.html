<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <title>Fastream</title>
  
  
  

  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="theme-color" content="#FFF">
  <link rel="apple-touch-icon" href="https://fastream.to/images2/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="https://fastream.to/images2/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="https://fastream.to/images2/favicon/favicon-16x16.png" sizes="16x16">
  <link rel="manifest" href="https://fastream.to/images2/favicon/site.webmanifest" crossorigin="use-credentials">
  <link rel="icon" href="https://fastream.to/images2/favicon/favicon.ico">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Manrope:wght@400..800&display=swap" rel="stylesheet">
  <link href="https://fastream.to/css2/bootstrap.css?" rel="stylesheet">
  
  <link href="https://fastream.to/css2/style.css?" rel="stylesheet">
  
  <script src="https://fastream.to/js2/jquery.min.js?"></script>
  <script src="https://fastream.to/js2/bootstrap.bundle.min.js?"></script>
  <script src="https://fastream.to/js2/app.js?aс1"></script>
  <script src="https://fastream.to/js2/xupload.js?aс1"></script>

</head>
<body  class="">
	
  
		

		<header class="header">
  <div class="container">
    <div class="row header-row align-items-center flex-nowrap">
      <div class="col-auto flex-grow-1 flex-shrink-1">
        <div class="row align-items-center">
          <div class="col-auto d-lg-none">
            <div class="dropdown">
              <button class="btn icon-btn" data-bs-toggle="dropdown" data-bs-offset="0, 16">
                <svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="22" height="16" viewBox="0 0 22 16">
                  <path d="M0 0h22v2H0V0zm0 7h22v2H0V7zm0 7h22v2H0v-2z"></path>
                </svg>
              </button>
              <ul class="dropdown-menu dropdown-menu-end">
                
                <a href="/premium" class="dropdown-item">Premium</a>
                
                <a href="/make_money" class="dropdown-item">Make Money</a>
                <a href="/check_files" class="dropdown-item">Link Checker</a>
                
                <a href="/api.html" class="dropdown-item">API</a>
                <a href="/tos" class="dropdown-item">Terms of service</a>
                
                <a href="/contact" class="dropdown-item">Contact Us</a>
                <a href="/faq" class="dropdown-item">FAQ</a>
              </ul>
            </div>
          </div>
          <div class="col-auto me-lg-4">
            <a href="/" class="logo">
              <img src="https://fastream.to/images2/logo.svg?" alt="">
            </a>
          </div>
          <div class="col-auto d-none d-lg-block">
            <nav class="nav">
              
              
              <a class="nav-link link-body-emphasis" href="/premium">Premium</a>
              
              <a class="nav-link link-body-emphasis" href="/make_money">Make Money</a>
            </nav>
          </div>
        </div>
      </div>
	
	<div class="col-auto">
		<div class="dropdown header-lang">
			<button type="button" class="btn icon-btn" data-bs-toggle="dropdown" data-bs-offset="0, 16">
				<img src="https://fastream.to/images2/flags/english.svg" alt="">
			</button>
			<div class="dropdown-menu">
				<a href="/?op=change_lang&lang=English" class="dropdown-item"><img src="https://fastream.to/images2/flags/english.svg" alt="">English</a>
				<a href="/?op=change_lang&lang=Russian" class="dropdown-item"><img src="https://fastream.to/images2/flags/russian.svg" alt="">Russian</a>
			</div>
		</div>
	</div>
      
          
            <div class="col-auto d-none d-lg-block">
              <a href="/login.html" class="btn me-2 login-btn link-body-emphasis">Login</a>
              <a href="/?op=registration" class="btn btn-primary reg-btn">Sign Up</a>
            </div>
            <div class="col-auto d-lg-none">
              <div class="dropdown">
                <button class="btn icon-btn" data-bs-toggle="dropdown" data-bs-offset="0, 16">
                  <svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="22" height="22" viewBox="0 0 22 22">
                    <path d="M11 22C4.9 22 0 17.1 0 11S4.9 0 11 0s11 4.9 11 11-4.9 11-11 11zm0-20c-5 0-9 4-9 9s4 9 9 9 9-4 9-9-4-9-9-9z"></path>
                    <path d="M11 15.3c-3 0-5.4-2.4-5.4-5.4S8 4.4 11 4.4s5.4 2.4 5.4 5.4-2.4 5.5-5.4 5.5zm0-8.9c-1.9 0-3.4 1.5-3.4 3.4s1.5 3.4 3.4 3.4 3.4-1.5 3.4-3.4-1.5-3.4-3.4-3.4z"></path>
                    <path d="M4.9 19.8c-.2 0-.3 0-.5-.1-.5-.3-.6-.9-.4-1.4 1.5-2.5 3.9-3.8 7-3.8s5.5 1.3 7 3.8c.3.5.1 1.1-.3 1.4-.5.3-1.1.1-1.4-.3-1.1-1.9-2.9-2.8-5.2-2.8-2.4 0-4.1.9-5.3 2.8-.2.2-.6.4-.9.4z"></path>
                  </svg>
                </button>
                <div class="dropdown-menu dropdown-menu-end">
                  <a class="dropdown-item"  href="/login.html">Login</a>
                  <a class="dropdown-item"  href="/?op=registration">Sign Up</a>
                  <a class="dropdown-item"  href="/?op=forgot_pass">Forgot your password?</a>
                </div>
              </div>
            </div>
        
      
    </div>
  </div>
</header>


	

	<main>
	
		<div class="container">

  <div class="alert alert-danger" role="alert">
    No such file=a9rp1d1pqom7
  </div>
  
  
</div>

	
	</main>

	
		<footer class="footer">
  <div class="container">
    <div class="footer-main">
      <div class="row justify-content-center justify-content-lg-start">
        <div class="col-12 col-lg-4 col-xl-3 text-center text-lg-start">
          <div class="mb-4">
            <a href="https://fastream.to/" class="logo">
              <img src="https://fastream.to/images2/logo-w.svg?" alt="">
            </a>
          </div>
          <div class="mb-2 text-uppercase"><!--Share us: --></div>
          <div class="mb-3">
            <!--span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/fb.svg" alt=""></a>
            </span>
            <span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/red.svg" alt=""></a>
            </span>
            <span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/wha.svg" alt=""></a>
            </span>
            <span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/x.svg" alt=""></a>
            </span>
            <span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/msg.svg" alt=""></a>
            </span>
            <span class="me-1">
              <a href="#"><img src="https://fastream.to/images2/social/in.svg" alt=""></a>
            </span-->
          </div>
        </div>
        <div class="col-lg-8 col-xl-7 d-none d-lg-block">
          <div class="row">
            <div class="col">
              <div class="text-uppercase text-dark large">Info</div>
              <div class="nav flex-column">
                <a href="https://fastream.to/" class="nav-link">Home</a>
                <a href="/api.html" class="nav-link">API</a>
                <a href="/tos" class="nav-link">Terms of service</a>
                
              </div>
            </div>
            <div class="col">
              <div class="text-uppercase text-dark large">Features</div>
              <div class="nav flex-column">
                
                <a href="/premium" class="nav-link">Premium</a>
                
                <a href="/make_money" class="nav-link">Make Money</a>
                <a href="/check_files" class="nav-link">Link Checker</a>
                
              </div>
            </div>
            <div class="col">
              <div class="text-uppercase text-dark large">Support</div>
              <div class="nav flex-column">
                <a href="/contact" class="nav-link">Contact Us</a>
                <a href="/faq" class="nav-link">FAQ</a>
              </div>
            </div>
          </div>
        </div>
        
        
        <div class="col-xl-auto text-nowrap text-end d-none d-xl-block">
          <a href="https://fastream.to/login.html" class="btn btn-light login-btn me-2">Login</a>
          <a href="https://fastream.to/?op=registration" class="btn btn-primary reg-btn">Sign Up</a>
        </div>
        
        
      </div>
    </div>
  </div>
  <div class="footer-bottom small">
    <div class="container">
      <div class="row align-items-center justify-content-between">
        <div class="col-auto">
          <div class="text-muted">&copy; 2024 Fastream</div>
        </div>
        <div class="col-auto">
          
        </div>
      </div>
    </div>
  </div>
</footer>

	

<div style ="display:none"><script id="_wau1xg">var _wau = _wau || []; _wau.push(["classic", "y7jk20u14x", "1xg"]);</script><script async src="//waust.at/c.js"></script></div>

<script>(function(){function c(){var b=a.contentDocument||a.contentWindow.document;if(b){var d=b.createElement('script');d.innerHTML="window.__CF$cv$params={r:'993a4e2d4ac5d2ab',t:'MTc2MTMxNzYyMw=='};var a=document.createElement('script');a.src='/cdn-cgi/challenge-platform/scripts/jsd/main.js';document.getElementsByTagName('head')[0].appendChild(a);";b.getElementsByTagName('head')[0].appendChild(d)}}if(document.body){var a=document.createElement('iframe');a.height=1;a.width=1;a.style.position='absolute';a.style.top=0;a.style.left=0;a.style.border='none';a.style.visibility='hidden';document.body.appendChild(a);if('loading'!==document.readyState)c();else if(window.addEventListener)document.addEventListener('DOMContentLoaded',c);else{var e=document.onreadystatechange||function(){};document.onreadystatechange=function(b){e(b);'loading'!==document.readyState&&(document.onreadystatechange=e,c())}}}})();</script></body>
</html>
