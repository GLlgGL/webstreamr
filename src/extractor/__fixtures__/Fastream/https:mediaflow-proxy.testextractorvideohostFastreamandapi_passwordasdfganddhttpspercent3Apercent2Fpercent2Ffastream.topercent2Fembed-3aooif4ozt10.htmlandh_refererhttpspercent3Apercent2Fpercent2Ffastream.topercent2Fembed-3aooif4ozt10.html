{"destination_url":"https://s28.fastream.to/hls2/09/00024/3aooif4ozt10_,l,n,h,x,.urlset/master.m3u8?t=nw6_aeN6pDROuXsJQwyrJVm3qHKFedVfbTNj5LgXH-4&s=1752863379&e=43200&v=1058373&i=0.3&sp=0","request_headers":{"user-agent":"Mozilla/5.0 (X11; Linux x86_64; rv:138.0) Gecko/20100101 Firefox/138.0","Accept":"*/*","Connection":"keep-alive","Accept-Language":"en-US,en;q=0.5","referer":"https://fastream.to/","origin":"https://fastream.to"},"mediaflow_proxy_url":"https://mediaflow-proxy.test/proxy/hls/manifest.m3u8","query_params":{"api_password":"asdfg"}}
