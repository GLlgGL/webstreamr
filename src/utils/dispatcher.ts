import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
