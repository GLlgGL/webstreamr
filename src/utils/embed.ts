import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
