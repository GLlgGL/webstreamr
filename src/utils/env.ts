import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
