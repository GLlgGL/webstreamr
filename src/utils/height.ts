import { Context } from '../types';
import { CustomRequestInit, Fetcher } from './Fetcher';

export const guessHeightFromPlaylist = async (ctx: Context, fetcher: Fetcher, url: URL, init?: CustomRequestInit): Promise<number | undefined> => {
  const m3u8Data = await fetcher.text(ctx, url, init);

  const heights = Array.from(m3u8Data.matchAll(/\d+x(\d+)|(\d+)p/g))
    .map(heightMatch => heightMatch[1] ?? heightMatch[2])
    .filter(height => height !== undefined)
    .map(height => parseInt(height));

  return heights.length ? Math.max(...heights) : /* istanbul ignore next */ undefined;
};
