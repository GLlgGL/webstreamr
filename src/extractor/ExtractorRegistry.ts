import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
