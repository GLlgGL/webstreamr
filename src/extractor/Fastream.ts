import bytes from 'bytes';
import { NotFoundError } from '../error';
import { Context, Format, Meta, UrlResult } from '../types';
import { buildMediaFlowProxyExtractorStreamUrl, supportsMediaFlowProxy } from '../utils';
import { Extractor } from './Extractor';

export class Fastream extends Extractor {
  public readonly id = 'fastream';

  public readonly label = 'Fastream (via MediaFlow Proxy)';

  public override viaMediaFlowProxy = true;

  public supports(ctx: Context, url: URL): boolean {
    return null !== url.host.match(/fastream/) && supportsMediaFlowProxy(ctx);
  }

  public override normalize(url: URL): URL {
    return new URL(url.href.replace('/e/', '/embed-').replace('/d/', '/embed-'));
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.href };

    const downloadUrl = new URL(url.href.replace('/embed-', '/d/'));
    const html = await this.fetcher.text(ctx, downloadUrl, { headers });

    if (/No such file/.test(html)) {
      throw new NotFoundError();
    }

    const playlistUrl = await buildMediaFlowProxyExtractorStreamUrl(ctx, this.fetcher, 'Fastream', url, headers);

    const heightAndSizeMatch = html.match(/\d{3,}x(\d{3,}), ([\d.]+ ?[GM]B)/) as string[];
    const titleMatch = html.match(/>Download (.*?)</) as string[];

    return [
      {
        url: playlistUrl,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          bytes: bytes.parse(heightAndSizeMatch[2] as string) as number,
          height: parseInt(heightAndSizeMatch[1] as string),
          title: titleMatch[1],
        },
      },
    ];
  };
}
