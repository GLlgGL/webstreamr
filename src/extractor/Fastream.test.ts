import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { Fastream } from './Fastream';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new Fastream(new FetcherMock(`${__dirname}/__fixtures__/Fastream`))]);

const ctx = createTestContext({ mediaFlowProxyUrl: 'https://mediaflow-proxy.test', mediaFlowProxyPassword: 'asdfg' });

describe('Fastream', () => {
  test('fastream.to embed', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://fastream.to/embed-3aooif4ozt10.html'))).toMatchSnapshot();
  });

  test('expired or deleted', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://fastream.to/embed-a9rp1d1pqom7.html'))).toMatchSnapshot();
  });
});
