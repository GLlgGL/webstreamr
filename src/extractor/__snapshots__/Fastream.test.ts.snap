// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Fastream expired or deleted 1`] = `[]`;

exports[`Fastream fastream.to embed 1`] = `
[
  {
    "format": "hls",
    "label": "Fastream (via MediaFlow Proxy)",
    "meta": {
      "bytes": 1073741824,
      "countryCodes": [],
      "height": 1080,
      "title": "black.mirror.s01e01.lat.720p.mp4",
    },
    "sourceId": "fastream_",
    "ttl": 900000,
    "url": "https://mediaflow-proxy.test/proxy/hls/manifest.m3u8?api_password=asdfg&h_user-agent=Mozilla%2F5.0+%28X11%3B+Linux+x86_64%3B+rv%3A138.0%29+Gecko%2F20100101+Firefox%2F138.0&h_Accept=*%2F*&h_Connection=keep-alive&h_Accept-Language=en-US%2Cen%3Bq%3D0.5&h_referer=https%3A%2F%2Ffastream.to%2F&h_origin=https%3A%2F%2Ffastream.to&d=https%3A%2F%2Fs28.fastream.to%2Fhls2%2F09%2F00024%2F3aooif4ozt10_%2Cl%2Cn%2Ch%2Cx%2C.urlset%2Fmaster.m3u8%3Ft%3Dnw6_aeN6pDROuXsJQwyrJVm3qHKFedVfbTNj5LgXH-4%26s%3D1752863379%26e%3D43200%26v%3D1058373%26i%3D0.3%26sp%3D0",
  },
]
`;
